models/Document.php

```diff
@@ -643,47 +643,77 @@ protected function doDelete(): void
 
     public function delete(): void
     {
-        $this->dispatchEvent(new DocumentEvent($this), DocumentEvents::PRE_DELETE);
-
-        $this->beginTransaction();
 
         try {
-            if ($this->getId() == 1) {
-                throw new Exception('root-node cannot be deleted');
-            }
+            $this->dispatchEvent(new DocumentEvent($this), DocumentEvents::PRE_DELETE);
+
+            // we wrap the save actions in a loop here, so that we can restart the database transactions in the case it fails
+            // if a transaction fails it gets restarted $maxRetries times, then the exception is thrown out
+            // this is especially useful to avoid problems with deadlocks in multi-threaded environments (forked workers, ...)
+            $maxRetries = 5;
+            for ($retries = 0; $retries < $maxRetries; $retries++) {
+
+                $this->beginTransaction();
+
+                try {
+                    if ($this->getId() == 1) {
+                        throw new Exception('root-node cannot be deleted');
+                    }
+
+                    $this->doDelete();
+                    $this->getDao()->delete();
 
-            $this->doDelete();
-            $this->getDao()->delete();
+                    $this->commit();
+
+                    //clear parent data from registry
+                    $parentCacheKey = self::getCacheKey($this->getParentId());
+                    if (RuntimeCache::isRegistered($parentCacheKey)) {
+                        /** @var Document $parent */
+                        $parent = RuntimeCache::get($parentCacheKey);
+                        if ($parent instanceof self) {
+                            $parent->setChildren(null);
+                        }
+                    }
+                    break; // transaction was successfully completed, so we cancel the loop here -> no restart required
+
+                } catch (Exception $e) {
+                    try {
+                        $this->rollBack();
+                    } catch (Exception $er) {
+                        // PDO adapter throws exceptions if rollback fails
+                        Logger::error((string)$er);
+                    }
 
-            $this->commit();
+                    // we try to start the transaction $maxRetries times again (deadlocks, ...)
+                    if ($e instanceof DeadlockException && $retries < ($maxRetries - 1)) {
```

Comment:
```suggestion
                    if ($e instanceof DeadlockException && $retries < ($maxRetries - 1)) {
```
should be OR `RetryableException` ? I've copied from the current implementation of the update methods and they are not consistent for every element on this

Reply:
Maybe it's time to create some sort of utility function/class for this retry pattern, since it's now implemented in various places?

Reply:
that's a nice input, thank you! 
i will refactor a little bit once ready

Reply:
@jdreesen  would refactoring to something like this be ideal in this situation? https://github.com/pimcore/pimcore/pull/17853/commits/275383a5274694b48d88f7c1a64e13d19477e4c4
if yes, i would apply to all element `save()` and dataobject `delete()`